### PR TITLE
[BUG] | [Performance Issues] | [MER-2649] | [MER-2245]

### PR DESCRIPTION
**Pull request by dtiwarATS**
This branch will fix the following tickets:

1. [BB: Viewing student attempts in review mode extremely slow.](https://eliterate.atlassian.net/browse/MER-2649)
2. [BB: Some students using Chromebooks in the BIO 100 CPP section are experiencing poor performance in Unit 8.](https://eliterate.atlassian.net/browse/MER-2245)

Hey @bsparks, could you please look at the PR and approve it.

I am still not sure why the "[Container build and push]" check is failing. Please let me know if I have to make any changes anywhere.

### CODE REVIEW

File: assets/src/adaptivity/rules-engine.ts [hunk 1]
```diff
@@ -66,11 +66,11 @@ const applyToEveryCondition = (top: TopLevelCondition | NestedCondition, callbac
   });
 };
 
-const evaluateValueExpression = (value: string, env: Environment) => {
+const evaluateValueExpression = (value: string, env: Environment, key?: string) => {
```
Comment (dtiwarATS) [inline assets/src/adaptivity/rules-engine.ts:69]: These changes were already approved as part of https://github.com/Simon-Initiative/oli-torus/pull/4309

File: assets/src/components/parts/janus-capi-iframe/ExternalActivity.tsx [hunk 2]
```diff
@@ -697,19 +683,33 @@ const ExternalActivity: React.FC<PartComponentProps<CapiIframeModel>> = (props)
...
-          //we don't want to evaluate a JSON string
-          const looksLikeJSON = looksLikeJson(value);
-          formatted[variable].value = looksLikeJSON
-            ? JSON.stringify(evaluateJsonObject(JSON.parse(value), scriptEnv))
-            : templatizeText(formatted[variable].value, simLife.snapshot, scriptEnv, true, false);
-        } else if (typeof value === 'string' && isMathExpr) {
-          if (value.search(/app\.|variables\.|stage\.|session\./) >= 0) {
-            // math expr could be like: 16^{\\frac{1}{2}}=\\sqrt {16}={\\editable{}} which does NOT need processing
-            formatted[variable].value = templatizeText(value, simLife.snapshot, scriptEnv, true);
+        if (context !== contexts.REVIEW) {
```
Comment (dtiwarATS) [inline assets/src/components/parts/janus-capi-iframe/ExternalActivity.tsx:686]: These changes were already approved as part of https://github.com/Simon-Initiative/oli-torus/pull/4309 (These changes were already approved as part of https://github.com/Simon-Initiative/oli-torus/pull/4309)

File: assets/src/adaptivity/rules-engine.ts [hunk 3]
```diff
@@ -347,33 +350,55 @@ export const getReferencedKeysInConditions = (conditions: any) => {
...
     }) => {
       // the fact *must* be a reference to a key we need
-      if (condition.fact) {
+      if (!checkExpressionInTarget && condition.fact) {
         references.add(condition.fact);
       }
       // the value *might* contain a reference to a key we need
       if (typeof condition.value === 'string') {
-        extractUniqueVariablesFromText(condition.value).forEach((v) => references.add(v));
+        const variablesInCondition = checkExpressionInTarget
```
Comment (dtiwarATS) [inline assets/src/adaptivity/rules-engine.ts:361]: To improve the performance, when a lesson is opened in authoring, we generate a list of variables (Init state, Rules and SIM - > Config Data) that contains expression and needs evaluation. We stored them in conditionsNeedEvaluation in activity.content.custom.conditionsNeedEvaluation.

File: assets/src/apps/delivery/store/features/groups/actions/deck.ts [hunk 4]
```diff
@@ -496,6 +496,15 @@ export const loadActivities = createAsyncThunk(
         hasMoreHints: result.hasMoreHints || true,
         groupId: null,
       };
+      //To improve the performance, when a lesson is opened in authoring, we generate a list of variables that contains expression and needs evaluation
+      // we stored them in conditionsNeedEvaluation in activity.content.custom.conditionsNeedEvaluation. When this function is called
+      // we only process variables that is present in conditionsNeedEvaluation array and ignore others.
```
Comment (bsparks) [inline assets/src/apps/delivery/store/features/groups/actions/deck.ts:501]: what is going to be done about things that haven't had this in the authoring tool? is there a migration needed?

**Review by bsparks — APPROVED** [review]

File: assets/src/apps/delivery/store/features/groups/actions/deck.ts (hunk 4 above)
Comment (dtiwarATS) [reply assets/src/apps/delivery/store/features/groups/actions/deck.ts:501]: I already mentioned that the lesson needs to be open once in the authoring tool in order this to work correctly. Apart from the Janus-text, other areas like Init State, Adaptive Rules, ConfigData is covered here.

**Comment by bitbldr** [comment]
Container build and push step is failing because the credentials are not present during the build step. We need to change the CI to conditionally run this step only if the credentials are present (like we do auto format). For now though, this is not a required check and therefore this PR should not be blocked by it, it is okay to proceed even though this failed

**Pull request merged**